### PR TITLE
[triton][inductor] Eliminate full-tile loads when the value repeats across the reduction (#192410)

### PR DESCRIPTION
**Pull request by rahul342**
Summary:

**Why:**
- An `ops.masked` reduction can make Inductor emit a full `[XBLOCK, R0_BLOCK]` load for a value that depends only on the output row. Every reduction lane then repeats address, predicate, and load work for the same per-row value.

- A representative graph is:

```python
def fn(index, source0, source1, value0, value1):
    gathered0 = torch.index_select(source0, 0, index)
    gathered1 = torch.index_select(source1, 0, index)
    values = torch.cat((value0 + gathered0, value1 + gathered1), dim=1)
    return values.square().sum(dim=-1)
```

- Elementwise, the first gather computes `gathered0[x, c, r] = source0[index[x], c, r]`. In the measured CMF instance, Inductor flattens `(row, concatenation position)` into `xindex` and recovers the row as `x1 = xindex // 192`:

```python
# Before: the same index[x] value is loaded in every reduction lane.
tmp7 = tl.load(in_ptr1 + tl.broadcast_to(x1, [XBLOCK, R0_BLOCK]), ...)

# After: load once per row and broadcast at the reduction consumer.
tmp7 = tl.load(in_ptr1 + tl.broadcast_to(x1, [XBLOCK, 1]), ...)
```

- The gathered-data loads and reduction remain `[XBLOCK, R0_BLOCK]`; only the invariant index load stays narrow. With `XBLOCK=4` and `R0_BLOCK=128`, the old Triton IR represents four logical index values as a `4x128` load in each masked branch. This is not a claim of 128x HBM traffic; it removes redundant lane-wise address, predicate, load, register, and dependent-wait work.

- In the artifact-derived CMF representative, an isolated A/B changing only these load shapes reduced latency from approximately 277 us to 191 us, registers from 68 to 50, and wait instructions from 49 to 17. This is mechanism evidence; model-level effects are reported below.

**What:**
- Use existing block-shape propagation to prove that a load address and every non-range predicate are invariant over one or more reduction axes, then emit the load at the smallest proven shape.
- Qualify the lane-shape change on CUDA/HIP with a non-throwing current-device check. Other Triton backends and missing-device codegen retain dense indexing.
- Keep unproven cases dense: unknown mask shapes, reduction-dependent addresses or predicates, fixed-tile reductions whose positive extent cannot be proven, cooperative and mix-order reductions, explicit dense indexing, block pointers/TMA, `copy_shape`, and `override_mask`.
- Looped reductions do not require a positive-extent proof because an empty `tl.range` executes no body. Dynamic combo subkernels can omit a pointwise axis because a zero-width PID interval executes no branch; an end-to-end zero-extent combo regression covers that dispatch path.
- This diff targets `load()`. D114995186 extends the same proof to `index_expr()` and `check_bounds()`; native-matmul indexing remains separate.
- Generated-code tests live in `test_indexing.py` and use the existing `run_and_get_kernels` / `FileCheck` utilities. Coverage includes looped and persistent reductions, direct and indirect loads, partial axes, dynamic combo zero-width dispatch, negative legality cases, and the distinct FP32, BF16-upcast, and boolean load-emission paths.

**Fixed-class benchmark, latency reduction:**

| **Class** | **MI300** | **MI350** | **H100** |
| -- | -- | -- | -- |
| Pointwise-axis omission | 2.00-2.18% | 2.28-2.33% | 2.65% |
| Combo kernel, one grid topology | 1.67% | Not timed; gates pass | 2.01%, 16/20 blocks, drift observed |

**mts_gpu_benchmark QPS:**

| **Model** | **MI300** | **MI350** | **H100** | **Verdict** |
| -- | -- | -- | -- | -- |
| CMF | +0.487% | +0.255% | +0.865% | Win |
| HIM Facebook | +0.087%, 90% CI [-0.023%, +0.198%] | +0.065%, one card | +0.164% | Neutral, inside the +/-0.25% band |
| OBA | No eligible site | No eligible site | Not timed | N/A |
| HIM Instagram | No eligible site | No eligible site | Not timed | N/A |

`No eligible site` means the control artifact had no masked reduction load matching the proof. Per-card provenance is in a comment on this diff.

Test Plan:
- Seven focused `TritonKernel.indexing()` legality tests pass, covering backend gating, looped and fixed-tile extents, unknown masks, schedule guards, `copy_shape`, and `override_mask`. TestX: https://www.internalfb.com/intern/testinfra/testrun/14636698978229753

```
buck test -c fbcode.disable_re_tests=True 'fbcode//caffe2/test/inductor:codegen_triton' -- test_reduction_invariant_load_indexing --env PYTORCH_TEST_REMOTE_GPU=0 --env HIP_VISIBLE_DEVICES=2 --timeout=600 --print-passing-details
```

- The dynamic sequential-combo zero-pointwise-extent regression passes on MI300. TestX: https://www.internalfb.com/intern/testinfra/testrun/31525197416972585

- The final D1 source passes all 19 `ReductionInvariantIndexingTests` on an idle MI300 GPU (20 Buck results including listing/aggregate; 0 failures/skips). TestX: https://www.internalfb.com/intern/testinfra/testrun/28710447649876688

```
buck test fbcode//mode/opt 'fbcode//caffe2/test/inductor:indexing' -- ReductionInvariantIndexingTests --env HIP_VISIBLE_DEVICES=2 --timeout=1200 --print-passing-details
```

- Standard and extra `arc lint` pass. Scoped production Pyrefly reports zero `bad-argument-type` errors, Citrine reports no applicable findings, and fbcode/xplat mirrors are byte-identical.

- The production mechanism and fixed-class timing matrices were exercised on MI300, MI350, and H100. `mts_gpu_benchmark` crossover runs produced the model table in the summary; accepted arms passed source, package, configuration, idle, and no-lowering/no-autotuning gates.

Differential Revision: D114905866

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review this

**Comment by rahul342** [comment]
workign on address claude review comments

**Comment by rahul342** [comment]
Addressing Claude's finding:

| Review finding | Resolution in this PR | Disposition |
|---|---|---|
| Dead conjunct in the `outside_loop_vars` guard | Simplified the condition to `not reduction_axes_omitted`, removed the now-unused `has_indirect` local, and updated the comment to describe the actual loop-local placement invariant. | Addressed |
| Mixed `str`/`int` block-shape comparison | Both sides are now normalized with `str()` before comparison. | Addressed |
| Indirect scalar-shape construction | Replaced the generator expression with `("1",) * self.triton_tensor_ndim()`. | Addressed |
| Abstract helper docstring | Added a concrete `[XBLOCK, R0_BLOCK]` to `[XBLOCK, 1]` example and explained where broadcasting occurs. | Addressed |
| Ambiguous `reduction_invariant_indexing` naming | Renamed the caller request to `allow_reduction_invariant_indexing` and the result field to `reduction_axes_omitted`. | Addressed |
| Adjacent `IndexingOptions` type checks | Folded them into one `isinstance` block. | Addressed |
| Undocumented CUDA-only gate | Kept the gate because only CUDA and ROCm have been qualified, and added a comment explaining that other Triton backends retain dense indexing until covered. | Addressed; backend restriction intentionally retained |
| Opt-in kwarg looked transitional | Kept it as an explicit consumer capability: callers opt in only when they can consume an indexing shape with omitted reduction axes. The request/result contract is now documented. | Retained and clarified |
| Generated-code tests were in `test_cuda_repro.py` | Moved them to a dedicated `ReductionInvariantIndexingTests` class in `test_indexing.py`. They remain accelerator-gated because the optimization is currently qualified only on CUDA/ROCm. | Addressed |
| Bespoke regex-based Triton parser | Removed the custom parser and switched to existing `run_and_get_kernels` / `run_and_get_code` helpers with `FileCheck`. | Addressed |
| Loop-placement test depended on indentation | Replaced indentation inspection with `FileCheck` ordering around the reduction loop, narrow load, reduction, and epilogue load. | Addressed |
| White-box template test only verified argument forwarding | Removed the template-specific forced opt-out. Narrowing now structurally requires `copy_shape is None`; templates preserve their explicitly owned shape through `copy_shape`. Added a direct test that an explicit `copy_shape` prevents narrowing. | Addressed with a structural rule rather than a compiled-template test |
| `no_x_dim` test manually created potentially unreachable state | The test now constructs a real `FixedTritonConfig` that naturally produces the `tensor_dim=None/0` mapping. The combo path also has generated-code coverage. Other low-level state patches remain for isolated legality guards. | Concrete reachability concern addressed |
| No explicit non-CUDA gate test | Not added to this PR. Non-CUDA backends remain dense by construction, while explicit backend-gate coverage is handled in the stacked consumer follow-up. | Follow-up |
| Looped reductions still re-emit the narrow load each chunk | Clarified the current loop-local behavior in the comment. Hoisting was not added: indexing-shape invariance alone does not prove that all address/predicate producers are safely hoistable, and hoisting can increase live ranges and register pressure. | Deferred as a separate optimization |


@claude review this again

**Comment by rahul342** [comment]
@claude one more review

File: test/inductor/test_codegen_triton.py [hunk 1]
```diff
@@ -195,6 +198,247 @@ def should_use_persistent_reduction(features, cooperative_reduction):
         self.assertFalse(kernel.persistent_reduction)
         self.assertEqual(seen_scores, [tiling_scores])
 
+    def test_reduction_invariant_load_indexing(self):
```
Comment (eellison) [inline test/inductor/test_codegen_triton.py:201]: nit: IMO, these type of invasive kernel tests are better as hooks or assertions on the codegen..but theyre fine..

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -4548,6 +4575,78 @@ def _broadcast_shape_with_masks(
...
+            if dim is None or str(shape[dim]) == str(dense_shape[dim]):
+                continue
+            if not tree.is_reduction:
+                if str(shape[dim]) != "1":
+                    return None
+                continue
+            if (
+                str(shape[dim]) != "1"
+                # A zero-length reduction makes the existing dense load fully masked.
+                or not V.graph.sizevars.statically_known_gt(tree.numel, sympy.S.Zero)
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: This also seemed suspect, comment with agent.

> Is the positive-extent requirement necessary for looped reductions? Narrowed loads are explicitly kept inside the reduction loop, so when `rnumel == 0`, `tl.range(0, rnumel, ...)` executes no iterations and no load occurs.
>
> This guard seems necessary only for persistent/non-loop reductions, where a fixed tile may execute despite a zero logical extent. Could this check be narrowed to:
>
> ```python
> not tree.is_loop and not V.graph.sizevars.statically_known_gt(
>     tree.numel, sympy.S.Zero
> )
> ```
>
> The extent unit test should then permit narrowing for unknown/zero looped reductions while retaining the restriction for persistent reductions.

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -4548,6 +4575,78 @@ def _broadcast_shape_with_masks(
...
+        # These schedules introduce coordinate scopes outside the block shape
+        # tracked by this proof.
+        if self.cooperative_reduction or self.mix_order_reduction:
+            return None
+
+        active_range_trees = self.active_range_trees()
+        # A non-combo kernel with a zero pointwise extent has an empty launch
+        # grid. A combo grid can still launch this subkernel because a sibling
+        # has more blocks, so require one valid pointwise coordinate here.
+        if self.is_combo_kernel and any(
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: This seemed suspect, and after discussion with agent - 

> I don’t think this combo-specific pointwise-extent guard is necessary. Dynamic combo kernels use cumulative PID ranges, so when a subkernel has `xnumel == 0`, its interval has zero width and no PID enters its body. Per-subkernel-block dispatch behaves similarly, while literal static-empty work is eliminated before combo formation.
>
> I reproduced the reachable case with unbacked `u0`/`u1` extents, bypassed only this guard, and ran `u0=0` alongside `u1=16`; the single combo kernel produced correct outputs with the narrowed load.
>
> Could we remove this guard and add an unbacked zero-extent combo regression test?

**Review by eellison — APPROVED** [review]
Looks good, although a couple of unnecessary comments.

**Comment by rahul342** [comment]
@jansel I did few rounds of claude reviews and corrections. PTAL.

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (rahul342) [reply torch/_inductor/codegen/triton.py]: addressing thsi

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (rahul342) [reply torch/_inductor/codegen/triton.py]: addressing this
